File: ruby/STO_Event_Calendar.rb
#!/usr/bin/env ruby

# frozen_string_literal: true

# rubocop:disable lint/UnneededCopEnableDirectve
# rubocop:disable Lint/MissingCopEnableDirective, Lint/UnneededCopDisableDirective, Metrics/LineLength

require 'date'
require 'optparse'
# rubocop:disable Layout/SpaceAroundOperators
require 'optparse/date'
# rubocop:enable Layout/SpaceAroundOperators
Stats = Struct.new(:needed, :claimed, :end, :daily, :reset)
stats = Stats.new

def gatherinfo(prompt)
  print(prompt)
  STDOUT.flush
  gets.chomp
end

# rubocop:disable Lint/UselessMethodArgument, Lint/UnusedMethodArgument
def print_dates(days, reset_time, rem, final)
  # rubocop:disable Style/StringLiterals
  puts "Estimated date of completion (if you get your tokens every day): #{days.strftime('%B %d')}"
  puts ""
  # rubocop:enable Style/StringLiterals
end
# rubocop:enable Lint/UselessMethodArgument

# rubocop:disable Metrics/LineLength, Style/DateTime
OptionParser.new do |parser|
  parser.on('-d', '--daily [DAILY]', Integer, 'Amount of tokens acquired per day.') { |d| stats.daily = d }
  parser.on('-n', '--needed [NEEDED]', Integer, 'Number of tokens needed to complete event.') { |n| stats.needed = n }
  parser.on('-c', '--claimed [CLAIMED]', Integer, 'Number of tokens already claimed.') { |c| stats.claimed = c }
  parser.on('-e', '--end [END]', DateTime, 'End date of the event.') { |e| stats.end = e }
  parser.on('-r', '--reset [RESET]', Float, 'Amount of hours until dailies reset.') { |r| stats.reset = r }
end
# rubocop:enable Metrics/LineLength

# rubocop:disable Metrics/LineLength, Style/StringLiterals
stats.daily = gatherinfo('Max tokens per day: ').to_i unless stats.daily.is_a?(Integer)
stats.needed = gatherinfo('Total tokens needed: ').to_i unless stats.needed.is_a?(Integer)
stats.claimed = gatherinfo('Tokens already claimed: ').to_i unless stats.claimed.is_a?(Integer)
stats.end = DateTime.strptime(gatherinfo('End Date (MM/DD/YYYY): '), "%m/%d/%Y") unless stats.end.respond_to?(:to_datetime)
stats.reset = gatherinfo('Hours until reset: ').to_f unless stats.reset.is_a?(Float)
# rubocop:enable Metrics/LineLength, Style/StringLiterals

today = DateTime.now
# rubocop:disable Lint/UselessAssignment
remaining = today - stats.end
days_needed = ((stats.needed - stats.claimed) / stats.daily).ceil
date_needed = today.to_date + days_needed.to_r if days_needed.respond_to?(:to_r)
final_day = stats.end - days_needed
puts "Max tokens per day: #{stats.daily}"
puts "Total tokens needed: #{stats.needed}"
puts "Number of tokens already claimed: #{stats.claimed}"
puts "End Date: #{stats.end}"
puts "Hours until reset: #{stats.reset}"
puts "Days remaining: #{remaining}"
puts "Today's Date: #{today}"
puts "Estimated date of completion: #{date_needed.to_time.strftime('%B %d')}"
puts "Number of days needed: #{days_needed}"
puts "Final day to be able to start the event: #{final_day.to_time.strftime('%B %d')}"

File: python/STO_Event_Calendar.py
#pylint: disable=invalid-name, missing-docstring, line-too-long
from __future__ import division, print_function
import sys

# try:
#     from __future__ import assertions  # putting this here in case i need it.
# except ImportError:
#     pass

import argparse

from datetime import datetime, timedelta
from math import ceil
if sys.version_info[0] >= 3:
    raw_input = input


def gatherinfo(string):
    '''This function prompts the user for input with the prompt being specified by the "string" argument.
       This function does not do any data verification, so any wrong data input will probably trigger a TypeError or ValueError later on.'''
    try:
        return raw_input(string)
    except NameError:
        return input(string)


parser = argparse.ArgumentParser(
    description="Calculator for Star Trek Online in-game events.")
parser.add_argument('--daily-tokens', '-d', type=int, dest='daily',
                    help='The amount of tokens you can get per day.')
parser.add_argument('--total-tokens', '-t', type=int, dest='total',
                    help='The total amount of tokens you need to complete the project.')
parser.add_argument('--tokens-claimed', '-c', type=int, dest='claimed',
                    help="The amount of tokens you've already claimed.")
parser.add_argument('--end-date', '-e', dest='end',
                    help="The date that the event ends.")
parser.add_argument('--daily-reset', '-r', dest='reset', type=float, default=20.0,
                    help="Time that the daily reset has left (in case you've already turned it in today), in hours.")
args = parser.parse_args()

# Gathering info about the current event

if args.daily:
    daily = args.daily
else:
    daily = int(gatherinfo("Max tokens per day: "))

if args.total:
    needed = args.total
else:
    needed = int(gatherinfo("Total tokens needed: "))

if args.claimed:
    tokens = args.claimed
else:
    tokens = int(gatherinfo("Tokens already claimed: "))

if args.end:
    end_date = args.end
else:
    end_date = gatherinfo("End date (MM/DD/YYY): ")

if args.reset:
    reset = args.reset
else:
    reset = float(gatherinfo("Hours until dailies reset: "))


# Calculating time left
td = datetime.today()
tod = td + timedelta(hours=reset)
end = datetime.strptime(end_date, "%m/%d/%Y")  # .date()
remaining = end - td
try:
    days_needed = timedelta(int(ceil((needed - tokens) / daily)))
    if days_needed.days < 0:
        raise ValueError("\nError: Days Needed has yielded a negative number.")
except ValueError as e:
    print(e)
    exit(1)
date_needed = td.date() + days_needed
final_day = end - days_needed


# The Output
# The Final Day represents the point of
# no return.  From that day forward, you must
# get tokens every day in order to complete
# the event and recieve the reward.

print('\nToday\'s date:', td.strftime("%B %d %Y"))
print('Days remaining in event:', remaining.days)


def print_dates():
    print("Estimated date of completion (if you get your tokens every day):",
          date_needed.strftime("%B %d"))
    print("Daily quest will be available at aproximately:",
          tod.strftime("%B %d %Y at %I:%M %p"))
    print('Days needed to complete event:', days_needed.days)


if final_day < td:
    print('The wormhole\'s closed, captain.')
    print('There\'s no way to get the reward for this event.')
elif final_day == td:
    print('You must claim the daily tokens every day from here on out.')
    print_dates()
else:
    print('Final day to start:', final_day.strftime("%B %d"))
    print_dates()
